LoopKit/DoseProgressTimerEstimator.swift

```diff
@@ -0,0 +1,96 @@
+//
+//  DoseProgressTimerEstimator.swift
+//  LoopKit
+//
+//  Created by Pete Schwamb on 3/23/19.
+//  Copyright © 2019 LoopKit Authors. All rights reserved.
+//
+
+import Foundation
+
+
+open class DoseProgressTimerEstimator: DoseProgressReporter {
+
+    private let lock = UnfairLock()
+
+    private var observers = WeakSet<DoseProgressObserver>()
+
+    var timer: DispatchSourceTimer?
+
+    let reportingQueue: DispatchQueue
+
+    public init(reportingQueue: DispatchQueue) {
+        self.reportingQueue = reportingQueue
+    }
+
+    open var progress: DoseProgress {
+        fatalError("progress must be implemented in subclasse")
+    }
+
+    public func addObserver(_ observer: DoseProgressObserver) {
+        lock.locked {
+            let firstObserver = observers.isEmpty
+            observers.insert(observer)
+            if firstObserver {
+                start()
+            }
+        }
+    }
+
+    public func removeObserver(_ observer: DoseProgressObserver) {
+        lock.locked {
+            observers.remove(observer)
+            if observers.isEmpty {
+                stop()
+            }
+        }
+    }
+
+    public func notify() {
+        let observersCopy = lock.locked { () -> WeakSet<DoseProgressObserver> in
+            return observers
+        }
+
+        for observer in observersCopy {
+            observer.doseProgressReporterDidUpdate(self)
+        }
+
+        if progress.isComplete {
+            lock.locked { stop() }
+        }
+    }
+
+    func start() {
+        guard self.timer == nil, !progress.isComplete else {
+            return
+        }
+
+        let (delay, repeating) = timerParameters()
+
+        let timer = DispatchSource.makeTimerSource(queue: reportingQueue)
+        timer.schedule(deadline: .now() + delay, repeating: repeating)
+        timer.setEventHandler(handler: { [weak self] in
+            self?.notify()
+        })
+        self.timer = timer
+        timer.resume()
+    }
+
+    open func timerParameters() -> (delay: TimeInterval, repeating: TimeInterval) {
+        fatalError("timerParameters must be been implemented in subclasse")
+    }
+
+    func stop() {
+        guard let timer = timer else {
+            return
+        }
+
+        timer.setEventHandler {}
```

Comment:
What's the effect of removing the event handler right before canceling the timer?

Reply:
Perhaps it is unnecessary, but I saw that usage here: https://medium.com/over-engineering/a-background-repeating-timer-in-swift-412cecfd2ef9, which covered some important details about setting up and deallocating GCD timers.